File: apps/sdkwork-im-h5/packages/sdkwork-im-h5-chat/src/services/chatTimelineUtils.ts
import type { SdkWorkListPageInfo, TimelineViewEntry } from "@sdkwork/im-sdk";

export const MAX_TIMELINE_ENTRIES = 500;

export interface TimelinePaginationState {
  hasMore: boolean;
  nextAfterSeq: number;
}

export function readCursorPageInfo(
  pageInfo?: SdkWorkListPageInfo,
): { hasMore: boolean; nextCursor?: string } {
  const hasMore = pageInfo?.hasMore === true;
  return {
    hasMore,
    nextCursor: hasMore ? (pageInfo?.nextCursor ?? undefined) : undefined,
  };
}

export function readSeqPageInfo(pageInfo?: SdkWorkListPageInfo): TimelinePaginationState {
  const hasMore = pageInfo?.hasMore === true;
  const parsed = hasMore && pageInfo?.nextCursor
    ? Number.parseInt(pageInfo.nextCursor, 10)
    : 0;
  return {
    hasMore,
    nextAfterSeq: Number.isFinite(parsed) && parsed > 0 ? parsed : 0,
  };
}

export function resolveLatestMessageSeq(entries: readonly TimelineViewEntry[]): number {
  return entries.reduce((max, entry) => Math.max(max, entry.messageSeq ?? 0), 0);
}

export function mergeTimelineEntries(
  existing: readonly TimelineViewEntry[],
  incoming: readonly TimelineViewEntry[],
): TimelineViewEntry[] {
  const byId = new Map<string, TimelineViewEntry>();
  for (const entry of existing) {
    byId.set(entry.messageId, entry);
  }
  for (const entry of incoming) {
    byId.set(entry.messageId, entry);
  }
  return Array.from(byId.values())
    .sort((left, right) => left.messageSeq - right.messageSeq)
    .slice(-MAX_TIMELINE_ENTRIES);
}

export function pickTimelinePagination(response: {
  pageInfo?: SdkWorkListPageInfo;
}): TimelinePaginationState {
  return readSeqPageInfo(response.pageInfo);
}

File: apps/sdkwork-im-flutter-mobile/packages/sdkwork_im_flutter_mobile_chat/lib/src/services/chat_timeline_utils.dart
import 'package:sdkwork_im_flutter_mobile_core/sdkwork_im_flutter_mobile_core.dart';

import 'chat_sdk_response_utils.dart';

/// Maximum timeline entries retained in memory per conversation (aligned with H5).
const int maxTimelineEntries = 500;

class ChatTimelineResult {
  const ChatTimelineResult({
    required this.items,
    required this.pagination,
  });

  final List<TimelineViewEntry> items;
  final TimelinePaginationState pagination;
}

class TimelinePaginationState {
  const TimelinePaginationState({
    required this.hasMore,
    required this.nextAfterSeq,
  });

  final bool hasMore;
  final int nextAfterSeq;
}

int resolveLatestMessageSeq(List<TimelineViewEntry> entries) {
  var maxSeq = 0;
  for (final entry in entries) {
    if (entry.messageSeq > maxSeq) {
      maxSeq = entry.messageSeq;
    }
  }
  return maxSeq;
}

List<TimelineViewEntry> mergeTimelineEntries(
  List<TimelineViewEntry> existing,
  List<TimelineViewEntry> incoming,
) {
  final byId = <String, TimelineViewEntry>{};
  for (final entry in existing) {
    byId[entry.messageId] = entry;
  }
  for (final entry in incoming) {
    byId[entry.messageId] = entry;
  }
  final merged = byId.values.toList()
    ..sort((left, right) => left.messageSeq.compareTo(right.messageSeq));
  if (merged.length <= maxTimelineEntries) {
    return merged;
  }
  return merged.sublist(merged.length - maxTimelineEntries);
}

TimelinePaginationState readSeqPageInfo(PageInfo? pageInfo) {
  final hasMore = pageInfo?.hasMore == true;
  final parsedCursor = hasMore ? int.tryParse(pageInfo?.nextCursor ?? '') : null;
  return TimelinePaginationState(
    hasMore: hasMore,
    nextAfterSeq: parsedCursor != null && parsedCursor > 0 ? parsedCursor : 0,
  );
}

TimelinePaginationState pickTimelinePagination(ChatTimelineResult? response) {
  return response?.pagination ??
      const TimelinePaginationState(hasMore: false, nextAfterSeq: 0);
}

ChatTimelineResult readTimelinePageFromSdkResponse(
  ConversationsMessagesListResponse? response,
) {
  final pageInfo = readPageInfoFromSdkData(response?.data);
  return ChatTimelineResult(
    items: readItemsFromSdkData(
      response?.data,
      TimelineViewEntry.fromJson,
    ),
    pagination: readSeqPageInfo(pageInfo),
  );
}

PostedMessageResponse? readPostedMessageFromSdkResponse(
  ConversationsMessagesCreateResponse201? response,
) {
  return readItemFromSdkData(response?.data, PostedMessageResponse.fromJson);
}
